src/tox/session/env_select.py

```diff
@@ -302,10 +321,11 @@ def _mark_active(self) -> None:
                 for env_info in self._defined_envs_.values():
                     if labels.intersection(env_info.env.conf["labels"]):
                         env_info.is_active = True
-            if self._state.conf.options.factors:  # if matches mark it active
+            if factors:  # if matches mark it active
                 for name, env_info in self._defined_envs_.items():
-                    if factors.issubset(set(name.split("-"))):
-                        env_info.is_active = True
+                    for factor_set in factors:
+                        if factor_set.issubset(set(name.split("-"))):
+                            env_info.is_active = True
```

Comment:
We should probably add here a `break` 🤔 as there's no point setting this more than once in any circumstance

Reply:
Yes, I noticed that about the previous logic as well -- I was thinking about adding an `if env_info.is_active: continue` to the top of the loop.

I'll add both optimizations: there's no point in continuing either loop when we know there's no possible change.